L1Trigger/L1TCaloLayer1/src/UCTLayer1.cc

```diff
@@ -0,0 +1,117 @@
+#include <iostream>
+#include <iomanip>
+#include <stdlib.h>
+#include <stdint.h>
+
+#include "UCTLayer1.hh"
+
+#include "UCTCrate.hh"
+#include "UCTCard.hh"
+#include "UCTRegion.hh"
+#include "UCTTower.hh"
+
+#include "UCTGeometry.hh"
+
+UCTLayer1::UCTLayer1() : uctSummary(0) {
+  UCTGeometry g;
+  for(uint32_t crate = 0; crate < g.getNCrates(); crate++) {
+    crates.push_back(new UCTCrate(crate));
+  }
+}
+
+UCTLayer1::~UCTLayer1() {
+  for(uint32_t i = 0; i < crates.size(); i++) {
+    if(crates[i] != 0) delete crates[i];
+  }
+}
+
+bool UCTLayer1::clearEvent() {
+  for(uint32_t i = 0; i < crates.size(); i++) {
+    if(crates[i] != 0) crates[i]->clearEvent();
+  }
+  return true;
+}
+
+const UCTRegion* UCTLayer1::getRegion(int regionEtaIndex, uint32_t regionPhiIndex) const {
+  if(regionEtaIndex == 0 || regionEtaIndex < -7 || regionEtaIndex > 7 ||
+     regionPhiIndex <= 0 || regionPhiIndex >= 19) {
+    return 0;
+  }
+  // Get (0,0) tower region information
+  UCTGeometry g;
+  UCTRegionIndex r = UCTRegionIndex(regionEtaIndex, regionPhiIndex);
+  UCTTowerIndex t = g.getUCTTowerIndex(r);
+  uint32_t absCaloEta = abs(t.first);
+  uint32_t absCaloPhi = abs(t.second);
+  uint32_t crt = g.getCrate(absCaloEta, absCaloPhi);
+  if(crt >= crates.size()) {
+    std::cerr << "UCTLayer1::getRegion - Crate number is wrong - " << std::hex << crt 
```

Comment:
> Speaking of, does CMSSW set a preprocessor flag?

I think CMSSW sets PROJECT_NAME to CMSSW and PROJECT_VERSION to e.g. CMSSW_8_0_3 .